torch/utils/data/datapipes/iter/combining.py

```diff
@@ -55,12 +58,106 @@ def __iter__(self):
 
 @functional_datapipe('fork')
 class ForkIterDataPipe(IterDataPipe):
+    r""" :class:`ForkIterDataPipe`.
+
+        Iterable DataPipe to create multiple instances of the same Iterable DataPipe.
+        args:
+            datapipe: Iterable DataPipe being copied
+            num_instances: number of instances of the datapipe to create
+            buffer_size: this restricts how far ahead the leading child DataPipe
+             can read relative to the slowest child DataPipe
+    """
+    def __new__(cls, datapipe: IterDataPipe, num_instances: int, buffer_size: int = 1000):
+        container = _ForkIterDataPipe(datapipe, num_instances, buffer_size)
+        return [ChildDataPipe(container, i) for i in range(num_instances)]
+
+    def __init__(self, *arg):
+        raise Exception("__init__ called instead of __new__")
```

Comment:
nit:
Do we need to split it into two classes? It seems we can combine them together.
```py
class Fork():
    def __new__(cls, dp, num_instances, buffer_size):
        container = super().__new__(cls)
        container.__init__(dp, num_instances, buffer_size)
        return [ChildDataPipe(container, i) for i in range(num_instances)]
    # Similar to `_ForkIterDataPipe`
    def __init__(self, dp, num_instances, buffer_size):
        ...
```
cc: @VitalyFedyunin

Reply:
Good suggestion, I think this works and I will try this out.

Reply:
That will work (and probably more readable)

Reply:
This may not work for the case of multiprocessing as @NivekT just pointed out. Serialization/Deserialization doesn't work as my expectation even with `__getnewargs__` implemented.
I suggest Kevin to keep it as previous. I will do some research on that.

Reply:
Did some research on the reason that it doesn't work well with pickling. Want to summarize my findings:
- When unpickling, __init__ is never called in most of cases
- If `__new__` method is specified, `__getnewargs__` is needed for pickler to pass the arguments to `__new__` method. As shown below, `attribute` in `load` function.
- My proposed method would work with `__new__` method that only returns a single element. But, since we are returning a tuple of elements, at the time of unpickling, the returned tuple of instances (`obj` in below) will be errored out due to missing `__dict__` attribute.
```py
def load(cls, attributes):
    obj = cls.__new__(cls)
    obj.__dict__.update(attributes)
    return obj
```

It turns out the `__new__` and `__init__` have to be splitted into two classes, if we want to enable the multiprocessing with `fork`. Sorry for the noise.